WordPress/src/main/java/org/wordpress/android/networking/menus/MenusRestWPCom.java

```diff
@@ -0,0 +1,162 @@
+package org.wordpress.android.networking.menus;
+
+import android.content.Context;
+import android.text.TextUtils;
+
+import com.android.volley.VolleyError;
+import com.wordpress.rest.RestRequest;
+
+import org.json.JSONArray;
+import org.json.JSONException;
+import org.json.JSONObject;
+import org.wordpress.android.WordPress;
+import org.wordpress.android.models.MenuModel;
+import org.wordpress.android.util.AppLog;
+
+import java.net.HttpURLConnection;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import static org.wordpress.android.util.VolleyUtils.statusCodeFromVolleyError;
+import static org.wordpress.android.networking.menus.MenusDataModeler.*;
+
+/**
+ * Interface for WordPress.com Menus REST API methods.
+ *
+ * ref: https://developer.wordpress.com/docs/api/
+ */
+public class MenusRestWPCom {
+    public interface IMenusDelegate {
+        Context getContext();
+        long getSiteId();
+        void onMenuReceived(int statusCode, MenuModel menu);
+        void onMenusReceived(int statusCode, List<MenuModel> menus);
+        void onMenuCreated(int statusCode, MenuModel menu);
+        void onMenuDeleted(int statusCode, MenuModel menu, boolean deleted);
+        void onMenuUpdated(int statusCode, MenuModel menu);
+    }
+
+    private static final String MENU_REST_PATH = "/sites/%s/menus/%s";
+    private static final String MENUS_REST_PATH = "/sites/%s/menus";
+    private static final String CREATE_MENU_REST_PATH = "/sites/%s/menus/new";
+    private static final String DELETE_MENU_REST_PATH = "/sites/%s/menus/%s/delete";
+
+    //
+    // JSON keys for fetchAllMenus response object
+    //
+    public static final String MENU_KEY = "menu";
+    public static final String ALL_MENUS_MENUS_KEY = "menus";
+
+    private static final String DELETED_KEY = "deleted";
+
+    private final IMenusDelegate mDelegate;
+
+    public MenusRestWPCom(IMenusDelegate delegate) {
+        if (delegate == null) {
+            throw new IllegalArgumentException("IMenusDelegate cannot be null");
+        }
+        mDelegate = delegate;
+    }
+
+    public boolean createMenu(final MenuModel menu) {
+        // a name string is required for the REST call
+        if (menu == null || TextUtils.isEmpty(menu.name)) return false;
+        String path = formatPath(CREATE_MENU_REST_PATH, null);
+        Map<String, String> params = new HashMap<>();
+        params.put(MENU_NAME_KEY, menu.name);
+        WordPress.getRestClientUtilsV1_1().post(path, params, null, new RestRequest.Listener() {
+            @Override public void onResponse(JSONObject response) {
+                menu.menuId = response.optLong(MENU_ID_KEY);
+                mDelegate.onMenuCreated(HttpURLConnection.HTTP_OK, menu);
```

Comment:
I see two things here: if we were meant to send the actual HTTP response status code, we should send what we got from the server (i.e. `201 created` for REST API compliance, or if WP.com API is not using 201 then at least whatever status code we actually got from the underlying layer) rather than hardcoding the `HttpURLConnection.HTTP_OK`, but second and most important, I think I wouldn't send the HTTP status code down to the listener at all. At this level of abstraction, the sole call to `onMenuCreated` should suffice to indicate success. On any other case, we should have an `onErrorResponse` or similar method to call, and a set of `enum`-defined errors to pass as a parameter, for instance `ERROR_MENU_NOT_CREATED`, `ERROR_MENU_ID_NOT_FOUND`, etc for each one of the possible errors the API is throwing. The docs in the API are not very good, but for instance if you try to fetch a menu with id = 0 you get a `400` status code and the following JSON body:
`{
  "error": "menu-id",
  "message": "Menu ID must be greater than 0."
}
`
I’m inclined to having a java representation in our code for this kind of errors.